FIP-0083: Add built-in Actor events in the Verified Registry, Miner and Market Actors (#872)
filecoin-project/FIPs: FIPS/fip-0083.md

---
fip: "0083"
title: Add built-in actor events
author: "Aarsh Shah (@aarshkshah1992)"
discussions-to: https://github.com/filecoin-project/FIPs/discussions/754
status: Draft
type: Technical (Core)
category: Core
created: 2023-11-27
---

## Summary

> This FIP defines and proposes the implementation of a meaningful subset of fire-and-forget externally observable events that will be emitted by the built-in Verified Registry, Storage Market and Storage Miner Actors. These enable external agents (henceforth referred to as *“clients”*) to observe and track a specific subset of on-chain data-cap allocation/claims, deal lifecycle and sector lifecycle state transitions.

Reviewer: ```suggestion
This FIP defines and proposes the implementation of a meaningful subset of fire-and-forget externally observable events that will be emitted by the built-in Verified Registry, Storage Market and Storage Miner Actors. These enable external agents (henceforth referred to as *“clients”*) to observe and track a specific subset of on-chain datacap allocation/claims, deal lifecycle and sector lifecycle state transitions.
```

## Abstract
[FIP 0049](https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0049.md) introduced and implemented the ability for actors to emit fire-and-forget externally observable events during execution.

This FIP aims to be the first of many for implementing events in the built-in actors. This specific FIP enumerates, defines and proposes the implementation of a subset of events for data-cap allocation/claims, deal lifecycle and sector life cycle transitions in the Verified Registry, Market and Miner Actors.

The events introduced in this FIP will enable *clients* such as network monitoring tools, block explorers, SP <> client deal brokering systems, browser clients, filecoin blockchain oracles etc. to rely on these events as a source of truth for sourcing information about the specific subset of on-chain activity and transitions captured by these events.

Events are emitted by the Verified Registry Actor for each verifier balance update, each allocation created, expired or claimed and for each claim that is updated or removed.

Events are emitted by the Market Actor for each deal that is published, activated, terminated or completed successfully.

> Events are emitted by the Miner Actor for each sector that is precommitted, activated, updated (SnapDeals) or terminated by an SP.

Reviewer: ```suggestion
Events are emitted by the Miner Actor for each sector that is precommitted, activated (provecommited), updated (SnapDeals) or terminated by an SP.
```

Filecoin network observability tools, block explorers, SP monitoring dashboards, deal brokering software, etc. need to rely on complex low-level techniques such as state tree diffs to source information about on-chain activity and state transitions. Emitting the select subset of built-in actor events defined in this FIP will make it easy for these clients to source the information they need by simply subscribing to the events and querying chain state.

## Specification
 _Please see [FIP-0049](https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0049.md) for a detailed explanation of the FVM event schema and the possible values each field_ 
 in the event schema can take. The interpretation of the flags field in each event entry is as follows:

 1. {"flags": "0x03”} ⇒ Index both key and value of the event entry
 2. {"flags": "0x02”} ⇒ Index only the value of the event entry
 3. {"flags": "0x01”} ⇒ Index only the key of the event entry

 _All values in the event payloads defined below are encoded with CBOR and so the codec field for each event_ 
 entry will be set to 0x51.  The codec field is left empty in the payloads defined below only for the 

>  sake of brevity.

Reviewer 2: I'm pretty sure we don't want to introduce "CBOR" (`0x51`) to the chain, I don't believe we use that anywhere at the moment and it would be a shame to add that to the mix. I think what we want here is "DAG-CBOR" (`0x71`).

The two problems with just plain CBOR are:

1. It's not strictly defined in any of our specs, we basically lean on the upstream CBOR specification which brings all sorts of wild additions (simple types, tagged extensions, lots of other things that are not so friendly to a content addressed and distributed world) and non-deterministic encoding forms.
2. It doesn't materialise links (CIDs) and a generic "CBOR" encoder typically will reject them, even though we managed to reserve the tag in the upstream tag registry.

Also since `0x51` is a rarely used codec in our stack, applications that consume this data will often not even have a codec registered to decode it, thereby adding an additional (tho minor) burden for consuming or validating this data.

### FVM support for CBOR encoding
The FVM currently only supports event values encoded with the `raw` encoding. This FIP adds
support for CBOR encoding (0x51), and built-in actor event values will be encoded with CBOR.

### Verified Registry Actor Events

#### Verifier Balance Updated
This event is emitted when the balance of a verifier is updated in the Verified Registry Actor.

The event payload is defined as:

| flags                | key | value                                |
|----------------------| --- |--------------------------------------|
| Index Key + Value    | “$type” | “verifier-balance” (string)          |
| Index Key + Value    | “verifier” | <VERIFIER_ACTOR_ID> (int)            |
| Index Key            | “balance” | <VERIFIER_DATACAP_BALANCE> (bigint) |

#### Datacap Allocated

> This event is emitted when a verified client allocates datacap to a specific data set and storage provider.  

Reviewer: ```suggestion
This event is emitted when a verified client allocates datacap to a specific data piece and storage provider.  
```

The event payload is defined as:

| flags | key | value                   |
| --- | --- |-------------------------|

> | Index Key + Value | “$type” | “allocation” (string)   |

Reviewer: out of curiosity, is there any reason why this can't just be an index key `allocation`?

Author: We want users to be able to quickly query for/filter for where events where (key="$type" AND value="allocation").

| Index Key + Value | “id” | <ALLOCATION_ID> (int)   |
| Index Key + Value | “client” | <CLIENT_ACTOR_ID> (int) |

> | Index Key + Value | “provider” | <SP_ACTOR_ID> (int)     |

Reviewer: peer: I think there is a value to also expose `amount` as a value in this event for datacap explorers - so they can get metrics like total datacap allocated by simply subscribing the events without having to query the allocation state

Reviewer: similarly, would be great if we can also expose the pieceCID, thinking of FIL+ bot and such tools can take advantage of it

Author: Captured in FIP discussion at https://github.com/filecoin-project/FIPs/discussions/754#discussioncomment-7991866.

#### Datacap Allocation Removed
This event is emitted when a datacap allocation that is past its expiration epoch is removed.

The event payload is defined as:

| flags | key | value                         |
| --- | --- |-------------------------------|
| Index Key + Value | “$type” | “allocation-removed” (string) |
| Index Key + Value | “id” | <ALLOCATION_ID> (int)         |
| Index Key + Value | “client” | <CLIENT_ACTOR_ID> (int)       |
| Index Key + Value | “provider” | <SP_ACTOR_ID> (int)           |

#### Allocation Claimed
This event is emitted when a client allocation is claimed by a storage provider after the corresponding verified data is provably committed to the chain.

The event payload is defined as:

| flags | key | value                 |
| --- | --- |-----------------------|
| Index Key + Value | “$type” | “claim” (string)      |
| Index Key + Value | “id” | <CLAIM_ID> (int)      |
| Index Key + Value | “client” | <CLIENT_ACTOR_ID> (int) |

> | Index Key + Value | “provider” | <SP_ACTOR_ID> (int)   |

Reviewer: similar suggestion on adding the amount

Author: Captured in FIP discussion at https://github.com/filecoin-project/FIPs/discussions/754#discussioncomment-7991866.

| Index Key + Value | “$type” | "deal-completed" (string) |
| Index Key + Value | “id” | <DEAL_ID> (int)           |
| Index Key + Value | "client" | <STORAGE_CLIENT_ACTOR_ID> (int)   |
| Index Key + Value | "provider" | <STORAGE_PROVIDER_ACTOR_ID> (int) |

### Miner Actor Events
The Miner Actor emits the following sector lifecycle events:

#### Sector Pre-committed
The sector pre-committed event is emitted for each new sector that is successfully pre-committed by an SP.

The event payload is defined as:

| flags | key | value                          |
| --- | --- |--------------------------------|
| Index Key + Value | “$type" | "sector-precommitted" (string) |
| Index Key + Value | “sector” | <SECTOR_NUMER> (int)           |

#### Sector Activated
This event is emitted for each pre-committed sector that is successfully activated by a storage provider. For now, sector activation corresponds 1:1 with prove-committing a sector but this can change in the future.

The event payload is defined as:

| flags | key | value                       |
| --- | --- |-----------------------------|
| Index Key + Value | “$type" | "sector-activated" (string) |

> | Index Key + Value | “sector” | <SECTOR_NUMER> (int)        |

Reviewer: I think we could use more information here for sector <> data monitoring. i.e: adding a `value` of `has-data` boolean; piece informations (cid, associated claim, market actor address)

(in general, the reason why we are adding these information is to ease the integration of data storage toolings, so unless there is a reason for limiting the info expose, i.e: costly gas (I dont think its going to be $$ tho?), size.. otherwise I think we should try to expose necessary info to enable toolings to be able to get necessary information by just subscribing the events without further message monitoring &state diff needed

Author: We are including more information to this event in https://github.com/filecoin-project/FIPs/pull/897 and https://github.com/filecoin-project/FIPs/discussions/754#discussioncomment-7861187.

This includes `piece-cid`, `piece size` and `unsealed cid`. Wdym by the `has-data` boolean ?

#### Sector Updated
This event is emitted for each CC sector that is updated to contained actual sealed data.

The event payload is defined as:

| flags | key | value                     |
| --- | --- |---------------------------|
| Index Key + Value | “$type" | "sector-updated" (string) |
| Index Key + Value | “sector” | <SECTOR_NUMER> (int)      |

#### Sector Terminated
The sector termination event is emitted for each sector that is marked as terminated by a storage provider. 

The event payload is defined as:

| flags | key | value                        |
| --- | --- |------------------------------|
| Index Key + Value | “$type" | "sector-terminated" (string) |
| Index Key + Value | “sector” | <SECTOR_NUMER> (int)         |

## Design Rationale
Richer and more detailed payloads for the events were initially considered. However, a large event payload 

> increases the event emission gas costs are borne by users of the built-in actor methods. The payload of 

Reviewer: how much gas are we talking about here?

Author: We don't expect this to be  significant but the exact amount here has yet to be profiled.

are terminated by the cron job and so the event is also emitted from the cron job. 
However, once [FIP-0074](https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0074.md) is implemented, 
this event will be emitted from the `OnMinerSectorsTerminate` method itself. This will make the event available to 
clients and the gas cost for this event will be paid by the `OnMinerSectorsTerminate` method.

The above is also true for the `deal-completed` event.  Deal completion is currently processed by a 
cron job and so the event will not be usable by clients.
[FIP-0074](https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0074.md) will ensure that the processing of completed deals is done as part of a method called by the 
Storage Provider thus making this event available to clients

### Miner Actor Events
All Miner Actor Events will already have the Miner Actor ID in the `emitter` field of the event.

So, we only need to include the `sector number` in the event payload for Miner Actor events to enable 
clients to filter Miner Actor events by the Miner/Sector they are interested in.

> An SP can activate a single sector by using the `ProveCommitSector` method. As of today, this method 

Reviewer: editor: adding a note about https://github.com/filecoin-project/FIPs/pull/820? depends the sequence of fip acceptances and finalizations, this note might be less relevant

However, this FIP does not propose changing the existing event schema or changing any of the chain 
data structures including the `MessageReceipt` structure. So, this change should be completely backward 
compatible.

## Test Cases
All pre-existing unit tests in the Verified Registry, Market and Miner Actors should expect and observe the emission of the events defined in this proposal wherever applicable.

Existing integration tests should also expect and observe the emission of events defined in this proposal wherever applicable.

## Security Considerations
This proposal does not introduce any reductions in security.

## Incentive Considerations
Even though the event payloads have been kept minimal to minimise the gas costs for emitting events, 
this proposal does increase the gas cost of methods in the Verified Registry, Miner and Market Actors that 
emit the events defined in this FIP.

While this increased gas cost is borne by the users of the methods, the benefit of the emitted events 
is enjoyed by networking monitoring tools, network accounting tools, block explorer tools and other 
external agents that provide some value-added services to the Filecoin network.  Improved 

> implementation of these tools should indirectly benefit the network users.

Reviewer: remove `should` or  -> `will`

Author: Done.
